host/src/lib.rs

```diff
@@ -0,0 +1,122 @@
+// Required for SP1
+#![feature(generic_const_exprs)]
+#![allow(incomplete_features)]
+
+// Copyright 2023 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub mod error;
+pub mod execution;
+pub mod metrics;
+pub mod preflight;
+pub mod provider_db;
+pub mod request;
+pub mod server;
+
+use std::{alloc, fmt::Debug, path::PathBuf};
+
+use anyhow::{Context, Result};
+use cap::Cap;
+use serde::{Deserialize, Serialize};
+use structopt::StructOpt;
+
+use crate::{error::HostError, request::ProofRequestOpt};
+
+#[global_allocator]
+static ALLOCATOR: Cap<alloc::System> = Cap::new(alloc::System, usize::MAX);
+
+#[derive(StructOpt, Default, Clone, Serialize, Deserialize, Debug)]
+#[serde(default)]
+pub struct Opt {
+    #[structopt(long, require_equals = true, default_value = "0.0.0.0:8080")]
+    /// Server bind address
+    /// [default: 0.0.0.0:8080]
+    address: String,
+
+    #[structopt(long, require_equals = true, default_value = "16")]
+    /// Limit the max number of in-flight requests
+    pub concurrency_limit: usize,
+
+    #[structopt(long, require_equals = true)]
+    pub log_path: Option<PathBuf>,
+
+    #[structopt(long, require_equals = true, default_value = "7")]
+    pub max_log: usize,
+
+    #[structopt(long, require_equals = true, default_value = "host/config/config.json")]
+    /// Path to a config file that includes sufficent json args to request
+    /// a proof of specified type. Curl json-rpc overrides its contents
+    config_path: PathBuf,
+
+    #[structopt(long, require_equals = true)]
+    /// Use a local directory as a cache for input. Accepts a custom directory.
+    cache_path: Option<PathBuf>,
+
+    #[structopt(long, require_equals = true, env = "RUST_LOG", default_value = "info")]
+    /// Set the log level
+    pub log_level: String,
+
+    #[structopt(flatten)]
+    /// Proof request options
+    pub proof_request_opt: ProofRequestOpt,
+}
+
+#[derive(Debug, Clone)]
+pub struct ProverState {
+    pub opts: Opt,
+}
+
+impl ProverState {
+    pub fn init() -> Result<Self, HostError> {
+        // Read the command line arguments;
+        let mut opts = Opt::from_args();
+        // Read the config file.
+        let mut file_config = ProofRequestOpt::from_file(&opts.config_path)?;
```

Comment:
So it looks like we cannot use the config anymore to store the non proof request `Opt` values? Only possible to define them currently using the command line? Not sure how important this really is, but the current settings file does support this. Could be a bit of a limitation going forward so my feeling is to make it possible because it may be pretty useful.

Reply:
We can support it with the same priority, I just wasn't aware of it's full usage currently.

Reply:
https://github.com/taikoxyz/raiko/pull/31/commits/9e558d7083ba60d7bd49d15487ffdfdece0ed2eb Here is how this could be done

Reply:
Just looking at the code so maybe I miss something, but does it allow for example `address` and `log_level` from `Opt` to be stored in the config file?

Reply:
https://github.com/taikoxyz/raiko/blob/9e558d7083ba60d7bd49d15487ffdfdece0ed2eb/host/src/lib.rs#L103 This function now reads the config file as if it was the `Opt` struct and merges the values

Reply:
Oh I see I see. Does the config needs to be updated so the proof request data is not in the root anymore?

Reply:
It does not need to be nested, but we can nest it if need be. It is just flattened via structop and serde to be collected more easily inside of the code.